File: app-runtime-management/agents/windows/src/TimeOnChrome.AppRuntime.Core/ApplicationClassification.cs
namespace TimeOnChrome.AppRuntime.Core;

public sealed record AppMatchCondition(string Field, string Value);
public sealed record AppMatchExpression(string Operator, IReadOnlyList<AppMatchCondition> Conditions);
public sealed record AppProductSelector(string Platform, AppMatchExpression Match);
public sealed record AppProduct(string Id, string Name, string Type, IReadOnlyList<AppProductSelector> Selectors);
public sealed record ApplicationDiscoverySummary(string Role, string NameSource, IReadOnlyList<string> SourceKinds,
    string? ObjectKind = null, string? ParentProductKey = null, string? VariantRole = null,
    string? Scope = null, string? SourceKind = null, string? EvidenceLevel = null,
    string? ApplicationOrigin = null, string? OriginEvidenceCode = null);
public sealed record AppEvidence(string Platform, string RuntimeIdentity, string DisplayName,
    IReadOnlyDictionary<string, string> Values, IReadOnlyList<string> VerifiedFields, string? ProductId = null,
    ApplicationDiscoverySummary? Discovery = null);
public sealed record AppClassificationRule(string Id, string Name, string Kind, AppMatchExpression Match,
    IReadOnlyList<AppMatchExpression> Exclude, string Mode, string Classification, string Type,
    bool Enabled, string Source, string Reason, string? Platform = null, string? ProductId = null);
public sealed record ChildProductClassification(string ProductId, string Classification);
public sealed record ChildApplicationBinding(string ChildId,
    IReadOnlyList<ChildProductClassification> Products, IReadOnlyList<string> RuleIds);
public sealed record ApplicationKnowledge(int SchemaVersion, long Version,
    IReadOnlyList<AppProduct> Products, IReadOnlyList<AppClassificationRule> Rules,
    IReadOnlyList<ChildApplicationBinding> Bindings);
public sealed record AppClassificationResolution(string? ProductId, string Classification,
    string Status, IReadOnlyList<string> RuleIds, IReadOnlyList<string> Suggestions,
    string AppType = "unknown", string TypeStatus = "unknown", string TypeReasonCode = "none");

public static class ApplicationClassifier
{
    private static readonly HashSet<string> Strong = new(StringComparer.Ordinal)
        { "runtimeIdentity", "binaryHash", "packageId", "distributionKey", "productKey", "hostedAppId", "signerKey" };
    public static bool SafeAutomatic(AppMatchExpression expression) =>
        expression.Conditions.Count > 0 && (expression.Operator == "all"
            ? expression.Conditions.Any(condition => Strong.Contains(condition.Field))
            : expression.Operator == "any" && expression.Conditions.All(condition => Strong.Contains(condition.Field)));

    public static bool Matches(AppMatchExpression expression, AppEvidence evidence, bool requireVerified = false)
    {
        bool Condition(AppMatchCondition condition)
        {
            var actual = condition.Field == "runtimeIdentity" ? evidence.RuntimeIdentity
                : evidence.Values.GetValueOrDefault(condition.Field);
            return string.Equals(actual, condition.Value, StringComparison.Ordinal)
                && (!requireVerified || !Strong.Contains(condition.Field)
                    || evidence.VerifiedFields.Contains(condition.Field, StringComparer.Ordinal));
        }
        return expression.Conditions.Count > 0 && (expression.Operator == "all"
            ? expression.Conditions.All(Condition)
            : expression.Operator == "any" && expression.Conditions.Any(Condition));
    }
    private static int Rank(string kind) => kind switch
    { "product" => 0, "family" or "developer" => 1, _ => 2 };

    public static (string? ProductId, string AppType, string Status, string ReasonCode) ResolveProductType(
        ApplicationKnowledge knowledge, AppEvidence evidence)
    {
        var products = knowledge.Products.Where(product => product.Selectors.Any(selector =>
            selector.Platform == evidence.Platform && SafeAutomatic(selector.Match)
                && Matches(selector.Match, evidence, true))).ToArray();
        if (products.Length != 1) return (null, "unknown", "unknown", "none");
        var product = products[0];
        var selector = product.Selectors.First(item => item.Platform == evidence.Platform
            && SafeAutomatic(item.Match) && Matches(item.Match, evidence, true));
        var fields = selector.Match.Conditions.Select(item => item.Field).ToHashSet(StringComparer.Ordinal);
        var reason = fields.Contains("distributionKey") ? "distributionProductRule"
            : fields.Contains("packageId") ? "exactPackageRule" : "verifiedProductRule";
        return (product.Id, product.Type, product.Type == "unknown" ? "unknown" : "confirmed",
            product.Type == "unknown" ? "none" : reason);
    }

    public static AppClassificationResolution Resolve(ApplicationKnowledge knowledge, string childId,
        AppEvidence evidence, string previous = "unclassified")
    {
        var identified = knowledge.Products.Where(product => product.Selectors.Any(selector =>
            selector.Platform == evidence.Platform && SafeAutomatic(selector.Match)
                && Matches(selector.Match, evidence, true))).Select(product => product.Id)
            .Order(StringComparer.Ordinal).ToArray();
        var productId = identified.Length == 1 ? identified[0] : null;
        var productType = ResolveProductType(knowledge, evidence);
        var binding = knowledge.Bindings.FirstOrDefault(item => item.ChildId == childId);
        var explicitEntry = binding?.Products.FirstOrDefault(item => item.ProductId == productId);
        if (identified.Length > 1) return new(null, previous, "conflict", [], [], productType.AppType, productType.Status, productType.ReasonCode);
        if (explicitEntry is not null) return new(productId, explicitEntry.Classification, "explicit", [], [], productType.AppType, productType.Status, productType.ReasonCode);
        var enabled = new HashSet<string>(binding?.RuleIds ?? [], StringComparer.Ordinal);
        var candidates = knowledge.Rules.Where(rule => rule.Enabled && enabled.Contains(rule.Id)
            && (rule.Platform is null || rule.Platform == evidence.Platform)
            && (rule.ProductId is null || rule.ProductId == productId)
            && (knowledge.SchemaVersion >= 2 && rule.Kind == "type" && rule.Match.Conditions.Count == 0
                ? productType.Status == "confirmed" && productType.AppType == rule.Type
                : rule.ProductId is not null && rule.Match.Conditions.Count == 0
                    || Matches(rule.Match, evidence, rule.Mode == "automatic"))
            && !rule.Exclude.Any(expression => Matches(expression, evidence))).ToArray();
        var suggestions = candidates.Where(rule => rule.Mode == "suggestion").Select(rule => rule.Id)
            .Order(StringComparer.Ordinal).ToArray();
        var automatic = candidates.Where(rule => rule.Mode == "automatic"
            && (knowledge.SchemaVersion >= 2 && rule.Kind == "type" && rule.Match.Conditions.Count == 0
                ? productType.Status == "confirmed" && productType.AppType != "unknown"
                : rule.ProductId is not null ? productId is not null : SafeAutomatic(rule.Match)))
            .OrderBy(rule => Rank(rule.Kind)).ThenBy(rule => rule.Id, StringComparer.Ordinal).ToArray();
        var best = automatic.Where(rule => automatic.Length > 0 && Rank(rule.Kind) == Rank(automatic[0].Kind)).ToArray();
        if (best.Length > 0)
        {
            var conflict = best.Select(rule => rule.Classification).Distinct(StringComparer.Ordinal).Count() > 1;
            return new(productId, conflict ? previous : best[0].Classification,
                conflict ? "conflict" : "automatic", best.Select(rule => rule.Id).ToArray(), suggestions,
                productType.AppType, productType.Status, productType.ReasonCode);
        }
        return new(productId, suggestions.Length > 0 ? previous : "unclassified", suggestions.Length > 0 ? "suggestion" : "unclassified", [], suggestions,
            productType.AppType, productType.Status, productType.ReasonCode);
    }
}

File: app-runtime-management/agents/macos/Sources/MacOSAppRuntimeCore/ApplicationClassification.swift
import Foundation

public struct AppMatchCondition: Codable { public let field: String; public let value: String }
public struct AppMatchExpression: Codable { public let `operator`: String; public let conditions: [AppMatchCondition] }
public struct AppProductSelector: Codable { public let platform: String; public let match: AppMatchExpression }
public struct AppProduct: Codable { public let id: String; public let name: String; public let type: String; public let selectors: [AppProductSelector] }
public struct ApplicationDiscoverySummary: Codable {
    public let role: String; public let nameSource: String; public let sourceKinds: [String]
    public init(role: String, nameSource: String, sourceKinds: [String]) {
        self.role = role; self.nameSource = nameSource; self.sourceKinds = sourceKinds
    }
}
public struct AppEvidence: Codable {
    public let platform: String
    public let runtimeIdentity: String
    public let displayName: String
    public let values: [String: String]
    public let verifiedFields: [String]
    public let productId: String?
    public let discovery: ApplicationDiscoverySummary?
    public init(platform: String, runtimeIdentity: String, displayName: String, values: [String: String],
                verifiedFields: [String], productId: String? = nil, discovery: ApplicationDiscoverySummary? = nil) {
        self.platform = platform; self.runtimeIdentity = runtimeIdentity; self.displayName = displayName
        self.values = values; self.verifiedFields = verifiedFields; self.productId = productId
        self.discovery = discovery
    }
}
public struct AppClassificationRule: Codable {
    public let id: String; public let name: String; public let kind: String
    public let platform: String?; public let productId: String?
    public let match: AppMatchExpression; public let exclude: [AppMatchExpression]
    public let mode: String; public let classification: String; public let type: String
    public let enabled: Bool; public let source: String; public let reason: String
}
public struct ChildProductClassification: Codable { public let productId: String; public let classification: String }
public struct ChildApplicationBinding: Codable {
    public let childId: String; public let products: [ChildProductClassification]; public let ruleIds: [String]
}
public struct ApplicationKnowledge: Codable {
    public let schemaVersion: Int; public let version: Int64
    public let products: [AppProduct]; public let rules: [AppClassificationRule]; public let bindings: [ChildApplicationBinding]
}
public struct AppClassificationResolution: Codable, Equatable {
    public let productId: String?; public let classification: String; public let status: String
    public let ruleIds: [String]; public let suggestions: [String]
}
public enum ApplicationClassifier {
    private static let strong: Set<String> = ["runtimeIdentity", "binaryHash", "packageId", "distributionKey", "signerKey"]
    public static func safeAutomatic(_ expression: AppMatchExpression) -> Bool {
        guard !expression.conditions.isEmpty else { return false }
        if expression.operator == "all" { return expression.conditions.contains { strong.contains($0.field) } }
        return expression.operator == "any" && expression.conditions.allSatisfy { strong.contains($0.field) }
    }
    public static func matches(_ expression: AppMatchExpression, _ evidence: AppEvidence, verified: Bool = false) -> Bool {
        func condition(_ item: AppMatchCondition) -> Bool {
            let actual = item.field == "runtimeIdentity" ? evidence.runtimeIdentity : evidence.values[item.field]
            return actual == item.value && (!verified || !strong.contains(item.field) || evidence.verifiedFields.contains(item.field))
        }
        guard !expression.conditions.isEmpty else { return false }
        if expression.operator == "all" { return expression.conditions.allSatisfy(condition) }
        return expression.operator == "any" && expression.conditions.contains(where: condition)
    }
    private static func rank(_ kind: String) -> Int {
        switch kind { case "product": return 0; case "family", "developer": return 1; default: return 2 }
    }
    public static func resolve(_ knowledge: ApplicationKnowledge, childId: String,
                               evidence: AppEvidence, previous: String = "unclassified") -> AppClassificationResolution {
        let identified = knowledge.products.filter { product in
            product.selectors.contains { $0.platform == evidence.platform && safeAutomatic($0.match) && matches($0.match, evidence, verified: true) }
        }.map(\.id).sorted()
        let productId = identified.count == 1 ? identified[0] : nil
        let binding = knowledge.bindings.first { $0.childId == childId }
        if identified.count > 1 { return .init(productId: nil, classification: previous, status: "conflict", ruleIds: [], suggestions: []) }
        if let explicit = binding?.products.first(where: { $0.productId == productId }) {
            return .init(productId: productId, classification: explicit.classification, status: "explicit", ruleIds: [], suggestions: [])
        }
        let enabled = Set(binding?.ruleIds ?? [])
        let candidates = knowledge.rules.filter { rule in
            rule.enabled && enabled.contains(rule.id) && (rule.platform == nil || rule.platform == evidence.platform)
                && (rule.productId == nil || rule.productId == productId)
                && ((rule.productId != nil && rule.match.conditions.isEmpty) || matches(rule.match, evidence, verified: rule.mode == "automatic"))
                && !rule.exclude.contains { matches($0, evidence) }
        }
        let suggestions = candidates.filter { $0.mode == "suggestion" }.map(\.id).sorted()
        let automatic = candidates.filter { $0.mode == "automatic" && ($0.productId != nil ? productId != nil : safeAutomatic($0.match)) }
            .sorted { rank($0.kind) == rank($1.kind) ? $0.id < $1.id : rank($0.kind) < rank($1.kind) }
        if let first = automatic.first {
            let best = automatic.filter { rank($0.kind) == rank(first.kind) }
            let conflict = Set(best.map(\.classification)).count > 1
            return .init(productId: productId, classification: conflict ? previous : first.classification,
                         status: conflict ? "conflict" : "automatic", ruleIds: best.map(\.id), suggestions: suggestions)
        }
        return .init(productId: productId, classification: suggestions.isEmpty ? "unclassified" : previous, status: suggestions.isEmpty ? "unclassified" : "suggestion",
                     ruleIds: [], suggestions: suggestions)
    }
}
